cli/tests/extensions.rs

```diff
@@ -0,0 +1,298 @@
+#[test]
+fn extension_list_should_emit_output() {
+    let tmp_dir = TmpDir::new();
+
+    // Output that no extension is installed when that is the case.
+    let cmd = Command::cargo_bin("phylum")
+        .unwrap()
+        .env("XDG_DATA_HOME", &tmp_dir)
+        .arg("extension")
+        .arg("list")
+        .assert();
+
+    let output = std::str::from_utf8(&cmd.get_output().stdout).unwrap();
+    let re = Regex::new(r#"No extension"#).unwrap();
+
+    assert!(output.lines().find(|m| re.is_match(m)).is_some());
+
+    // Install one extension
+    Command::cargo_bin("phylum")
+        .unwrap()
+        .env("XDG_DATA_HOME", &tmp_dir)
+        .arg("extension")
+        .arg("add")
+        .arg(fixtures_path().join("sample-extension"))
+        .assert();
+
+    // Output name and description of the extension when one is installed
+    let cmd = Command::cargo_bin("phylum")
+        .unwrap()
+        .env("XDG_DATA_HOME", &tmp_dir)
+        .arg("extension")
+        .arg("list")
+        .assert();
+
+    let output = std::str::from_utf8(&cmd.get_output().stdout).unwrap();
+    let re = Regex::new(r#"^sample-extension\s+This extension does a thing"#).unwrap();
+
+    assert!(output.lines().find(|m| re.is_match(m)).is_some());
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Miscellaneous tests
+////////////////////////////////////////////////////////////////////////////////
+
+#[test]
+fn valid_extension_is_loaded_correctly() {
+    let ext = Extension::try_from(fixtures_path().join("sample-extension")).unwrap();
+
+    assert_eq!(ext.name(), "sample-extension");
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Utilities
+////////////////////////////////////////////////////////////////////////////////
+
+fn project_root() -> PathBuf {
+    Path::new(&env!("CARGO_MANIFEST_DIR"))
+        .ancestors()
+        .nth(1)
+        .unwrap()
+        .to_path_buf()
+}
+
+fn fixtures_path() -> PathBuf {
+    project_root()
+        .join("cli")
+        .join("tests")
+        .join("fixtures")
+        .join("extensions")
+}
+
+struct TmpDir(PathBuf);
+
+impl TmpDir {
+    fn new() -> Self {
+        let dir_name: String = rand::thread_rng()
+            .sample_iter(&rand::distributions::Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
+
+        let path = tmp_path().join(dir_name);
+        std::fs::create_dir_all(&path).unwrap();
+        Self(path)
+    }
+}
+
+impl AsRef<Path> for TmpDir {
+    fn as_ref(&self) -> &Path {
+        &self.0
+    }
+}
+
+impl AsRef<OsStr> for TmpDir {
+    fn as_ref(&self) -> &OsStr {
+        &self.0.as_os_str()
+    }
+}
+
+impl Drop for TmpDir {
+    fn drop(&mut self) {
+        // std::fs::remove_dir_all(&self.0).unwrap();
+    }
+}
+
+fn tmp_path() -> PathBuf {
+    project_root().join("target").join("tests-tmp")
+}
```

Comment:
I would think a more natural temporary directory would be `std::env::temp_dir()`. Is there some reason that wouldn't work?

Reply:
That would most likely work, I went for this solution in the interest of manual inspections and holding everything under the repo (as `target` is meant to be "ephemeral" anyway). I'm open to switching it out if needs be.

Reply:
If you're going to put it in a system directory you might want to look at the tempfile crate. It has support for creating directories with prefixes, automatically cleaning up, and it also handles some permissions problems you can run into with temp directories, although I don't know if that's something we need to worry about when running these tests.